### PR TITLE
METRON-1589: '/api/v1/search/search' fails when 'Solr Zookeeper Urls' has comma separated multiple zookeeper urls

### PR DESCRIPTION
**Pull request by justinleet**
## Contributor Comments
Refactored `SolrDao` to kill `getZkHost` and made it `getZkHosts` and return a `List`. `getSolrClient` now takes a `List<String>`, and passes it down.

I also added a `getSolrClient` method that's just blank, and defers to the `getZkHosts` method.  Our calls to `getSolrClient` are generally `solrDao.getSolrClient(solrDao.getZkHosts)`, which seems unnecessary. I did not however, change everything to use that, but I could definitely be persuaded to as part of this.

Additionally, `SolrColumnMetadataDao` now takes a` SolrClient` instead of a Zookeeper host string, to make it consistent with the other DAOs.

There's one catch however: I haven't integration tested this, because I'd need multiple Zk nodes and we don't really have a clean way to get it (neither the `SolrComponent` or `ZkComponent` really let you do it).  I'm not sure if we want or need those tests, or if alternatively we want to spin up a multinode environment and manually test, or if we're fine just assuming things line up (given that it's really testing if the underlying method works as expected at this point). There are unit tests agains the methods to ensure splitting is handled properly in the methods.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrDao.java [hunk 1]
```diff
@@ -134,13 +134,30 @@ public void patch(RetrieveLatestDao retrieveLatestDao, PatchRequest request,
...
+
+  /**
+   * Get ZK hosts from the global config.
+   * @return List of ZkHosts
+   */
+  public List<String> getZkHosts() {
     Map<String, Object> globalConfig = accessConfig.getGlobalConfigSupplier().get();
-    return (String) globalConfig.get("solr.zookeeper");
+    String solrZookeeper = (String) globalConfig.get("solr.zookeeper");
+    return Arrays.asList(solrZookeeper.split(","));
```
Comment (cestella) [inline metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrDao.java]: Have you tried this when people put a space between the nodes?  Perhaps something like `Lists.newArrayList(Iterables.transform(Splitter.on(",").split(globalConfig.getOrDefault("solr.zookeeper", ""), x -> x.trim()));`


Also, `solr.zookeeper` should be a constant.

**Comment by merrimanr** [comment]
I think `solrDao.getSolrClient()` is easier to understand and if I had to choose I would use that.  Very minor though.

For integration testing I think it will take a non-trivial amount of work to test with a cluster of zk components.  I would be willing to compromise and leave this to multinode testing outside of our CI process.  As far as I know we don't test other components in this manner (ES for example).

**Comment by justinleet** [comment]
@cestella Check out the adjustment for the trim and see how you feel about it.

Where should solr.zookeeper live?  I was a going to drop it in SolrConstants, but discovered it currently lives in SolrWriter, which doesn't feel like the right place for it since it's more widely used.  Do we want to drop it into SolrConstants, and adjust SolrWriter to use that instead?

**Comment by cestella** [comment]
That looks good @justinleet Yeah, I think that key deserves to be in SolrConstants.

+1 by inspection, pending moving the constant.

**Comment by justinleet** [comment]
@cestella Before I merge this in, are you okay with this PR doing this for the existing constant?

https://github.com/apache/metron/pull/1040/commits/c04b9462669bbe26de0ee1f6c7d0d39a348e865b#diff-47dd38427181750f7a4d42b813b28f43R57

I wanted to keep the existing stuff consistent and avoid too much digging into something not entirely related.

Having said that, it seems like we could probably clean up some of configuration constants in a more holistic way, but I'd argue that's outside the scope of this PR.

File: metron-platform/metron-solr/src/main/java/org/apache/metron/solr/writer/SolrWriter.java [hunk 2]
```diff
@@ -17,47 +17,46 @@
...
-
-import static org.apache.metron.solr.SolrConstants.SOLR_WRITER_NAME;
-
 public class SolrWriter implements BulkMessageWriter<JSONObject>, Serializable {
 
   public static final String JAVA_SECURITY_CONFIG_PROPERTY = "java.security.auth.login.config";
 
   public enum SolrProperties {
-    ZOOKEEPER_QUORUM("solr.zookeeper"),
+    ZOOKEEPER_QUORUM(SolrConstants.SOLR_ZOOKEEPER),
```
Comment (cestella) [inline metron-platform/metron-solr/src/main/java/org/apache/metron/solr/writer/SolrWriter.java]: Yep, this is fine.  +1 by inspection.

**Comment by justinleet** [comment]
Closing PR that's in the feature branch

**Pull request closed without merging**